kimchi/src/circuits/gate.rs

```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
 /// analogously using `joint_combiner`.
 ///
 /// This function computes that combined value.
-pub fn combine_table_entry<'a, F: Field, I: DoubleEndedIterator<Item = &'a F>>(
+pub fn combine_table_entry<
+    'a,
+    F: 'a + Zero + One + Add + Mul + Clone,
+    I: DoubleEndedIterator<Item = &'a F>,
+>(
     joint_combiner: F,
     v: I,
 ) -> F {
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
 }
 
-impl<F: Field> SingleLookup<F> {
+impl<F: Copy> SingleLookup<F> {
     /// Evaluate the linear combination specifying the lookup value to a field element.
-    pub fn evaluate<G: Fn(LocalPosition) -> F>(&self, eval: G) -> F {
+    pub fn evaluate<K: Zero + Add + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
+        &self,
+        eval: G,
+    ) -> K {
         self.value
             .iter()
-            .fold(F::zero(), |acc, (c, p)| acc + *c * eval(*p))
+            .fold(K::zero(), |acc, (c, p)| acc + eval(*p) * *c)
     }
 }
 
 /// A spec for checking that the given vector belongs to a vector-valued lookup table.
 #[derive(Clone, Serialize, Deserialize)]
-pub struct JointLookup<F> {
-    pub table_id: usize,
-    pub entry: Vec<SingleLookup<F>>,
+pub struct JointLookup<SingleLookup> {
+    pub table_id: i32,
+    pub entry: Vec<SingleLookup>,
```

Comment:
why not just make this `Vec<SingleLookup<F>>` and the struct generic in `F`?

Reply:
Because `reduce` erases `SingleLookup` and returns a concrete value/`Expr` for the lookup

Reply:
yeah I see that now. So it's like we have multiple types. Do you think it'd be a good idea to define them somewhere and use them? Something like that:

```rust
type EvaluatedJointLookup<F> = JointLookup<F>;
type ExprSingleLookup<F> = SingleLookup<ConstantExpr<F>>;
type ExprJointLookup<F> = JointLookup<SingleLookup<ConstantExpr<F>>>;
```

might be a stupid idea

Reply:
We never use any of those types directly, not sure it's worth it.